### ISSUE TITLE
Crtl + Space in Template Editor does not work

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-xtend#1018]
Crtl + Space in Template Editor does not work

```
   [main] WARN  org.eclipse.xtext.ui.editor.contentassist.AbstractJavaBasedContentProposalProvider  - Error in polymorphic dispatcher : java.lang.IllegalStateException: No Root Parser Rule found; The Node Model is broken.
java.lang.RuntimeException: java.lang.IllegalStateException: No Root Parser Rule found; The Node Model is broken.
	at org.eclipse.xtext.ui.codetemplates.ui.projectedEditing.TemporaryResourceProvider.useTemporaryResource(TemporaryResourceProvider.java:40)
	at org.eclipse.xtext.ui.codetemplates.ui.contentassist.CodetemplatesProposalProvider.complete_Literal(CodetemplatesProposalProvider.java:313)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.xtext.util.PolymorphicDispatcher.invoke(PolymorphicDispatcher.java:303)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractJavaBasedContentProposalProvider.invokeMethod(AbstractJavaBasedContentProposalProvider.java:301)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractJavaBasedContentProposalProvider.completeRuleCall(AbstractJavaBasedContentProposalProvider.java:197)
	at org.eclipse.xtext.ui.codetemplates.ui.contentassist.AbstractCodetemplatesProposalProvider.completeTemplateBody_Parts(AbstractCodetemplatesProposalProvider.java:47)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.xtext.util.PolymorphicDispatcher.invoke(PolymorphicDispatcher.java:303)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractJavaBasedContentProposalProvider.invokeMethod(AbstractJavaBasedContentProposalProvider.java:301)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractJavaBasedContentProposalProvider.completeAssignment(AbstractJavaBasedContentProposalProvider.java:206)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractContentProposalProvider$DefaultContentAssistProcessorSwitch.caseAssignment(AbstractContentProposalProvider.java:67)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractContentProposalProvider$DefaultContentAssistProcessorSwitch.caseAssignment(AbstractContentProposalProvider.java:1)
	at org.eclipse.xtext.util.XtextSwitch.doSwitch(XtextSwitch.java:144)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractContentProposalProvider$DefaultContentAssistProcessorSwitch.accept(AbstractContentProposalProvider.java:73)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractContentProposalProvider.createProposals(AbstractContentProposalProvider.java:105)
	at org.eclipse.xtext.ui.editor.contentassist.AbstractJavaBasedContentProposalProvider.createProposals(AbstractJavaBasedContentProposalProvider.java:276)
	at org.eclipse.xtext.ui.editor.contentassist.CompletionProposalComputer.exec(CompletionProposalComputer.java:53)
	at org.eclipse.xtext.ui.editor.contentassist.CompletionProposalComputer.exec(CompletionProposalComputer.java:1)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:92)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:525)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.priorityReadOnly(XtextDocument.java:489)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.priorityReadOnly(XtextDocument.java:145)
	at org.eclipse.xtext.ui.editor.contentassist.RepeatedContentAssistProcessor.computeCompletionProposals(RepeatedContentAssistProcessor.java:118)
	at org.eclipse.xtext.ui.editor.contentassist.RepeatedContentAssistProcessor.computeCompletionProposals(RepeatedContentAssistProcessor.java:87)
	at org.eclipse.jface.text.contentassist.ContentAssistant$2.lambda$0(ContentAssistant.java:2015)
	at java.util.Collections$SingletonSet.forEach(Collections.java:4767)
	at org.eclipse.jface.text.contentassist.ContentAssistant$2.run(ContentAssistant.java:2014)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.text.contentassist.ContentAssistant.computeCompletionProposals(ContentAssistant.java:2011)
	at org.eclipse.jface.text.contentassist.CompletionProposalPopup.computeProposals(CompletionProposalPopup.java:576)
	at org.eclipse.jface.text.contentassist.CompletionProposalPopup.lambda$0(CompletionProposalPopup.java:506)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:72)
	at org.eclipse.jface.text.contentassist.CompletionProposalPopup.showProposals(CompletionProposalPopup.java:501)
	at org.eclipse.jface.text.contentassist.ContentAssistant.showPossibleCompletions(ContentAssistant.java:1825)
	at org.eclipse.jface.text.source.SourceViewer.doOperation(SourceViewer.java:978)
	at org.eclipse.jface.text.source.projection.ProjectionViewer.doOperation(ProjectionViewer.java:1383)
	at org.eclipse.xtext.ui.editor.embedded.TextViewerOperationAction$1.run(TextViewerOperationAction.java:125)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:72)
	at org.eclipse.xtext.ui.editor.embedded.TextViewerOperationAction.run(TextViewerOperationAction.java:122)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.commands.ActionHandler.execute(ActionHandler.java:121)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:97)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:318)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:252)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:156)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:487)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(KeyBindingDispatcher.java:308)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(KeyBindingDispatcher.java:584)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(KeyBindingDispatcher.java:653)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(KeyBindingDispatcher.java:443)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.access$2(KeyBindingDispatcher.java:386)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(KeyBindingDispatcher.java:96)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.filterEvent(Display.java:1846)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1422)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1449)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1432)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1473)
	at org.eclipse.swt.widgets.Widget.gtk_key_press_event(Widget.java:837)
	at org.eclipse.swt.widgets.Control.gtk_key_press_event(Control.java:3942)
	at org.eclipse.swt.widgets.Composite.gtk_key_press_event(Composite.java:855)
	at org.eclipse.swt.widgets.Widget.windowProc(Widget.java:2267)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:6679)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5928)
	at org.eclipse.swt.internal.gtk.GTK.gtk_main_do_event(Native Method)
	at org.eclipse.swt.widgets.Display.eventProc(Display.java:1486)
	at org.eclipse.swt.internal.gtk.OS.g_main_context_iteration(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4446)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.xtext.ui.codetemplates.ui.preferences.AdvancedTemplatesPreferencePage.editTemplate(AdvancedTemplatesPreferencePage.java:56)
	at org.eclipse.ui.texteditor.templates.TemplatePreferencePage.edit(TemplatePreferencePage.java:1202)
	at org.eclipse.ui.texteditor.templates.TemplatePreferencePage.edit(TemplatePreferencePage.java:1197)
	at org.eclipse.ui.texteditor.templates.TemplatePreferencePage.lambda$4(TemplatePreferencePage.java:926)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5687)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1423)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4955)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4448)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ui.internal.OpenPreferencesAction.run(OpenPreferencesAction.java:66)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:579)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:413)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5687)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1423)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4955)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4448)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:658)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:557)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:154)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:150)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:657)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:594)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1447)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1420)
Caused by: java.lang.IllegalStateException: No Root Parser Rule found; The Node Model is broken.
	at org.eclipse.xtext.nodemodel.util.NodeModelUtils.getEntryParserRule(NodeModelUtils.java:444)
	at org.eclipse.xtext.resource.XtextResource.getEntryPoint(XtextResource.java:436)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.initializeNodeAndModelData(ContentAssistContextFactory.java:198)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.initializeFromViewerAndResource(ContentAssistContextFactory.java:191)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.doCreateContexts(ContentAssistContextFactory.java:143)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.create(ContentAssistContextFactory.java:131)
	at org.eclipse.xtext.ui.editor.contentassist.antlr.DelegatingContentAssistContextFactory$StatefulFactory.doCreateContexts(DelegatingContentAssistContextFactory.java:93)
	at org.eclipse.xtext.ui.editor.contentassist.antlr.DelegatingContentAssistContextFactory$StatefulFactory.create(DelegatingContentAssistContextFactory.java:79)
	at org.eclipse.xtext.ui.editor.contentassist.antlr.DelegatingContentAssistContextFactory.create(DelegatingContentAssistContextFactory.java:56)
	at org.eclipse.xtext.ui.codetemplates.ui.contentassist.CodetemplatesProposalProvider$2.process(CodetemplatesProposalProvider.java:323)
	at org.eclipse.xtext.ui.codetemplates.ui.contentassist.CodetemplatesProposalProvider$2.process(CodetemplatesProposalProvider.java:1)
	at org.eclipse.xtext.util.concurrent.IUnitOfWork$Void.exec(IUnitOfWork.java:38)
	at org.eclipse.xtext.ui.codetemplates.ui.projectedEditing.TemporaryResourceProvider.useTemporaryResource(TemporaryResourceProvider.java:38)
	... 128 more
```

**Comment by cdietrich** [comment]
maybe we need to traverse down actions in a loop like

```
	public static ParserRule getEntryParserRule(INode node) {
		ICompositeNode root = node.getRootNode();
		EObject ge1 = root.getGrammarElement();
		if (ge1 instanceof ParserRule) {
			return (ParserRule) ge1;
		} else if (ge1 instanceof Action) {
			INode firstChild = root.getFirstChild();
			while (firstChild.getGrammarElement() instanceof Action && firstChild instanceof CompositeNode) {
				firstChild = ((CompositeNode)firstChild).getFirstChild();
			}
			EObject ge2 = firstChild.getGrammarElement();
			if (ge2 instanceof ParserRule) {
				return (ParserRule) ge2;
			}
		}
		throw new IllegalStateException("No Root Parser Rule found; The Node Model is broken.");
	}
```

@szarnekow any idea?

**Comment by szarnekow** [comment]
> maybe we need to traverse down actions in a loop like

Yes, the first element can be an action again.
That would be the case for syntaxes like 
```
ProductionRule:
  {P1} id=ID? {P2.p=current} string=STRING? {P3.p=current} i=INT;
```
Is that what you have in your grammar?

**Comment by cdietrich** [comment]
its the Xtend grammar ;)

```
Member returns XtendMember:
	{XtendMember} annotations+=XAnnotation*
	(
		(
			{XtendField.annotationInfo = current} modifiers+=CommonModifier*
```

**Comment by szarnekow** [comment]
:)

**Comment by cdietrich** [comment]
Fixed in 2.22
